site/cds_rdm/inspire_harvester/transform_entry.py

```diff
@@ -218,72 +447,230 @@ def _transform_alternate_identifiers(self):
+                )
+                return {}
+        return mapped
+
+    def _transform_experiments(self, inspire_experiments):
+        """Map experiments to CDS-RDM vocabulary."""
+        mapped = []
+        for experiment in inspire_experiments:
+            result = self._search_vocabulary(experiment, "experiments")
+            if result.get("hits", {}).get("total"):
+                mapped.append({"id": result["hits"]["hits"][0]["id"]})
+            else:
+                current_app.logger.warning(
+                    f"Couldn't map experiment '{experiment}' value to anything in existing vocabulary. INSPIRE record id: {self.inspire_metadata.get('control_number')}."
+                )
+                return {}
+        return mapped
+
+    def _transform_thesis(self, thesis_info):
+        """Transform thesis information to custom field format."""
+        defense_date = thesis_info.get("defense_date")
+        return {
+            "date_defended": defense_date,
+        }
+
     def transform_custom_fields(self):
         """Mapping of custom fields."""
         custom_fields = {}
         # TODO parse legacy name or check with Micha if they can expose name
-        accelerators = [
+        inspire_accelerators = [
             x.get("accelerator")
             for x in self.inspire_metadata.get("accelerator_experiments", [])
             if x.get("accelerator")
         ]
-        experiments = [
+        inspire_experiments = [
             x.get("legacy_name")
             for x in self.inspire_metadata.get("accelerator_experiments", [])
             if x.get("legacy_name")
         ]
 
-        custom_fields["cern:accelerators"] = accelerators
-        custom_fields["cern:experiments"] = experiments
+        thesis_info = self.inspire_metadata.get("thesis_info", {})
+        defense_date = thesis_info.get("defense_date")
+        if defense_date:
+            custom_fields["thesis:thesis"] = self._transform_thesis(thesis_info)
+
+        imprint = self._validate_imprint()
+
+        if inspire_accelerators:
+            mapped_accelerators = self._transform_accelerators(inspire_accelerators)
+            if mapped_accelerators:
+                custom_fields["cern:accelerators"] = mapped_accelerators
+        if inspire_experiments:
+            mapped_experiments = self._transform_experiments(inspire_experiments)
+            if mapped_experiments:
+                custom_fields["cern:experiments"] = mapped_experiments
+        if imprint and imprint.get("place"):
+            custom_fields["imprint:imprint"] = {"place": imprint.get("place")}
+
+        imprint_fields = custom_fields.get("imprint:imprint", {})
+
+        isbns = self.inspire_metadata.get("isbns", [])
+        online_isbns = []
+        for isbn in isbns:
+            value = isbn.get("value")
+            valid_isbn = normalize_isbn(value)
```

Comment:
this is a repetition, we can just use the already normalized ISBN, no need to normalise twice

Reply:
RIght, it is, but unless we cache it, I don't see how we can avoid it, as it's used in different places, for the record metadata and custom fields, which are the same level